addons/mail/tests/test_link_preview.py

```diff
@@ -76,13 +110,48 @@ def test_02_link_preview_create(self):
                             'id': link_preview.id,
                             'message': {'id': message.id},
                             'image_mimetype': False,
-                            'og_description': 'Test description',
-                            'og_image': False,
+                            'og_description': self.og_description,
+                            'og_image': self.og_image,
                             'og_mimetype': False,
-                            'og_title': 'Test title',
+                            'og_title': self.og_title,
                             'og_type': False,
-                            'source_url': 'https://thisdomainedoentexist.nothing',
+                            'source_url': self.source_url,
                         }] for link_preview in message.link_preview_ids
                     }
                 }]
             )
+
+    def test_get_link_preview_from_url(self):
+        test_cases = [
+            (self._patch_with_og_properties, self.source_url),
+            (self._patch_without_og_properties, self.source_url),
+            (self._patch_with_image_mimetype, self.og_image),
+        ]
+        test_asserts = [
+            {
+                'og_description': self.og_description,
+                'og_image': self.og_image,
+                'og_mimetype': None,
+                'og_title': self.og_title,
+                'og_type': None,
+                'source_url': self.source_url,
+            },
+            {
+                'og_description': None,
+                'og_image': None,
+                'og_mimetype': None,
+                'og_title': self.title,
+                'og_type': None,
+                'source_url': self.source_url,
+            },
+            {
+                'image_mimetype': 'image/png',
+                'og_image': self.og_image,
+                'source_url': self.og_image,
+            },
+        ]
+        session = requests.Session()
+        for (get_patch, url), expected in zip(test_cases, test_asserts):
+            with patch.object(requests.Session, 'get', get_patch):
+                preview = link_preview.get_link_preview_from_url(url, session)
```

Comment:
You could use subTests so that it does not halt at first issue

Reply:
Do you mean like so ? :slightly_smiling_face: 
```python
        session = requests.Session()
        for (get_patch, url), expected in zip(test_cases, test_asserts):
            with self.subTest(get_patch=get_patch, url=url, expected=expected):
                with patch.object(requests.Session, 'get', get_patch):
                    preview = link_preview.get_link_preview_from_url(url, session)
                    self.assertEqual(preview, expected)
```
I didn't set the 2 `with` on the same line this way if there's a problem with the `get_patch` it will also be catched by the subTest :thinking:

Reply:
Yes, indeed :) isn't the order of "with" doing the same as nesting them ?

Reply:
You're right :slightly_smiling_face: Thank you!